### ISSUE TITLE
if serviceaccount is deleted eksctl would not re-create it

### ISSUE THREAD

**Issue by pkit** [weaveworks/eksctl#4941]
**What were you trying to accomplish?**

```shell
eksctl create iamserviceaccount --override-existing-serviceaccounts --approve -f cluster.yaml
```

**What happened?**
```shell
2022-03-13 12:09:52 [ℹ]  no tasks
```
Although serviceaccount does not exist at all.

**How to reproduce it?**
1. Use `eksctl create iamserviceaccount` to create a serviceaccount
2. Use `kubectl delete` to delete serviceaccount
3. Run `eksctl create iamserviceaccount` again to see the "no tasks" and serviceaccount is still lost.

**Logs**

**Anything else we need to know?**
There is a workaround: delete the cloudformation stack that was used to create the serviceaccount.
Then `eksctl create iamserviceaccount` will re-create it.

**Versions**

```
$ eksctl info
eksctl version: 0.76.0
kubectl version: v1.22.4
OS: linux
```

**Comment by Skarlso** [comment]
Hi @pkit!

Interesting. Do you see this line by any chance in the log output?

```
... checking whether serviceaccount ...
```

**Comment by Skarlso** [comment]
@pkit Never mind. I found the problem. Now, the question is.... is it deliberate or are we just missing a pass in over the override flag.

**Comment by Skarlso** [comment]
We filter our existing names here: https://github.com/weaveworks/eksctl/blob/main/pkg/ctl/cmdutils/filter/iamserviceaccount_filter.go#L48

Note that the actually comment on the function is incorrect too, hehe. So this filters out existing names prior to even attempting to re-create them. I think this just misses the Plan ( aka override ) flag passed in. So, in case of an override, just ignore this filter I guess or something like that. Because, inside the code, we actually already have code that re-creates the service account if it doesn't exist. But that's never called because the name is already filtered out before hand anyways.

In any case, we'll prioritise this issue for the next available iteration. Thank you for reporting! We, of course, accept any kind of contribution if you wish this to be resolved sooner. :))

**Comment by pkit** [comment]
I really do not understand the logic that you have there. :)

**Comment by Skarlso** [comment]
It's basically just checking if it exists. And if it does, it will be ignored. That why you see something like this in the logs:

```
2022-03-14 07:15:18 [ℹ]  1 existing iamserviceaccount(s) (default/test-overwrite) will be excluded
2022-03-14 07:15:18 [ℹ]  1 iamserviceaccount (default/test-overwrite) was excluded (based on the include/exclude rules)
```

**Comment by Skarlso** [comment]
It checks if it exists in kubernetes.

**Comment by pkit** [comment]
> It checks if it exists in kubernetes.

But it doesn't.
Another scenario: same problem.
If I don't have permissions to create sa, the sa is not created, but the cf task stays there and any new re-run would not do anything too.

**Comment by Skarlso** [comment]
Interesting. But the error is definitely coming from there. So okay, did you see the above two lines in the log output?

**Comment by pkit** [comment]
> So okay, did you see the above two lines in the log output?

Yup, it's immediately excluded. In the first log line.
It's never checked in k8s, because I don't see any API calls

**Comment by Skarlso** [comment]
Oh I see what's happening. So we only do that if override is false, right? And it isn't false. So it got all the existing accounts from the stack which they do, right? Because the stack is there. And then it proceeded to filter them out in the call here:

```go
	return f.doSetExcludeExistingFilter(existing, "iamserviceaccount")
```

which does this:

```go
func (f *Filter) doSetExcludeExistingFilter(names []string, resource string) error {
	uniqueNames := sets.NewString(names...).List()
	f.excludeNames.Insert(uniqueNames...)
	for _, n := range uniqueNames {
		isAlsoIncluded := f.includeNames.Has(n) || f.matchGlobs(n, f.includeGlobs)
		if isAlsoIncluded {
			return fmt.Errorf("existing %s %q should be excluded, but matches include filter: %s", resource, n, f.describeIncludeRules())
		}
	}
	if len(uniqueNames) != 0 {
		logger.Info("%d existing %s(s) (%s) will be excluded", len(uniqueNames), resource, strings.Join(uniqueNames, ","))
	}
	return nil
}
```

Which is responsible for the log output. Now, we just have to find where the exclude is coming form.

**Comment by Skarlso** [comment]
It's added to the exclude list `	f.excludeNames.Insert(uniqueNames...)` in that function which is then used further down the line somewhere.

**Comment by Skarlso** [comment]
This guy filters them out:

```go
	filteredServiceAccounts := saFilter.FilterMatching(cfg.IAM.ServiceAccounts)
	saFilter.LogInfo(cfg.IAM.ServiceAccounts)
```

It's really not obvious what the heck is going on in there. :D

**Comment by Skarlso** [comment]
Note that this is linking the nodegroups, but these flags are also defined on the IAM account. But I couldn't find them in the docs. :)

**Comment by pkit** [comment]
I think I've got it, should be something like this:
```diff
diff --git a/pkg/ctl/cmdutils/filter/iamserviceaccount_filter.go b/pkg/ctl/cmdutils/filter/iamserviceaccount_filter.go
index 2f7e8ff8..8ffb23c7 100644
--- a/pkg/ctl/cmdutils/filter/iamserviceaccount_filter.go
+++ b/pkg/ctl/cmdutils/filter/iamserviceaccount_filter.go
@@ -55,23 +55,32 @@ func (f *IAMServiceAccountFilter) SetExcludeExistingFilter(stackManager serviceA
                return err
        }
 
-       if !overrideExistingServiceAccounts {
-               err := f.ForEach(serviceAccounts, func(_ int, sa *api.ClusterIAMServiceAccount) error {
-                       if api.IsEnabled(sa.RoleOnly) {
-                               return nil
-                       }
-                       exists, err := kubernetes.CheckServiceAccountExists(clientSet, sa.ClusterIAMMeta.AsObjectMeta())
-                       if err != nil {
-                               return err
-                       }
-                       if exists {
-                               existing = append(existing, sa.NameString())
-                       }
+       err = f.ForEach(serviceAccounts, func(_ int, sa *api.ClusterIAMServiceAccount) error {
+               if api.IsEnabled(sa.RoleOnly) {
                        return nil
-               })
+               }
+               exists, err := kubernetes.CheckServiceAccountExists(clientSet, sa.ClusterIAMMeta.AsObjectMeta())
                if err != nil {
                        return err
                }
+               if exists && !overrideExistingServiceAccounts {
+                       existing = append(existing, sa.NameString())
+               } else if !exists {
+                       found := -1
+                       for i := range existing {
+                               if existing[i] == sa.NameString() {
+                                       found = i
+                                       break
+                               }
+                       }
+                       if found >= 0 {
+                               existing = append(existing[:found], existing[found+1:]...)
+                       }
+               }
+               return nil
+       })
+       if err != nil {
+               return err
        }
 
        return f.doSetExcludeExistingFilter(existing, "iamserviceaccount")
```
The problem is that `existing` is always fintered out, but in case of a stale CF task it's should not be filtered.
That's why here I always do `ForEach` on service accounts and remove the ones that not exist from the `existing` slice.

**Comment by pkit** [comment]
Looks like `NewTasksToCreateIAMServiceAccounts` interface needs to be altered to support another set of SAs for "create in k8s only". Right now it will add a new CF task unconditionally.

**Comment by Skarlso** [comment]
We'll look at this during next planning which is tomorrow and try to come up with something. Several tests will have to be altered as well and / or fixed. :)

**Comment by Himangini** [comment]
We will try to reproduce this issue with the following steps provided, except the 2. step, we will use eksctl to delete the service account. If we can do this successfully then suggest the same solution to the user. 
```
1. Use eksctl create iamserviceaccount to create a serviceaccount
2. Use kubectl delete to delete serviceaccount
3. Run eksctl create iamserviceaccount again to see the "no tasks" and serviceaccount is still lost.
```

**Comment by pkit** [comment]
@Himangini I'm not sure how's that the same situation though
Service account can get deleted by deleting DS or Deployment that uses it.
It cannot be fully managed by `eksctl`

**Comment by tanvp112** [comment]
The service account created by `eksctl create iamserviceaccount` will not get deleted when you delete the daemonset/deployment that uses it.

**Comment by Skarlso** [comment]
This issue was already reproduced by me. We know are just debating a solution.

Gini was just summarising some approaches.

**Comment by Skarlso** [comment]
To summarise, and sorry for the confusion from all of you. :))) 

The problem is that if the service account is removed from kubernetes through none clean ways ( such as 'kubectl delete' rather than 'eksctl delete iamserviceaccount' which makes sure everything is deleted properly) the same `create` command 
 will not work, because it's detecting the stack for it and the filter always filters even if override is enabled or disabled.

As Gini is putting it nicely, it's possible to just delete everything and create it anew. That will clean up the stack and any remaining dangling objects. And it should be noted that you shouldn't just delete stuff from kubernetes randomly. :))

What I'm exploring, which is a nice UX feature,  is that much less of a hassle and takes less time to just run `create` again and re-create the same Kubernetes object without having to go through all the delete and create procedures again which takes time. But this requires a bit of thought on how to work precisely.

**Comment by pkit** [comment]
> The service account created by `eksctl create iamserviceaccount` will not get deleted when you delete the daemonset/deployment that uses it. The service account will get deleted when the underlying namespace was deleted, or explicitly delete via `kubectl`, however, the role on IAM will remain. The clean way to delete is `eksctl delete iamserviceaccount`.

As the namespace is not managed by `eksctl` it's not possible to reconcile k8s changes through gitops tools (`fluxcd`, `argocd`, etc.) and `eksctl` here.
I.e. removing `Namespace` through CD will bring `ekctl` to a broken state which cannot be recovered by rolling back git commits.

**Comment by Skarlso** [comment]
Okay, so we discussed this with the team.

There are two things that will happen. The first, is sadly, we aren't going to change how create works, because `eksctl` by default is not like `apply`. It's not aware of the environment in general, and changing a single command to be aware, will create an inconsistency between how the rest of the commands work. So for now the "solution" to this is to re-create.

Second. Though we do have initiative to change things, and we are working on an `apply` in general, so this has been added to that list. :) I can't promise though when that will happen. Probably not in the near future. :/

Sorry, I hope that this is satisfactory.
